core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+	s1 := math.cos_f32(1.0 / 16.0 * math.PI) / 2.0
+	s2 := math.cos_f32(2.0 / 16.0 * math.PI) / 2.0
+	s3 := math.cos_f32(3.0 / 16.0 * math.PI) / 2.0
+	s4 := math.cos_f32(4.0 / 16.0 * math.PI) / 2.0
+	s5 := math.cos_f32(5.0 / 16.0 * math.PI) / 2.0
+	s6 := math.cos_f32(6.0 / 16.0 * math.PI) / 2.0
+	s7 := math.cos_f32(7.0 / 16.0 * math.PI) / 2.0
+
+	if .info in options {
+		options += {.return_metadata, .do_not_decompress_image}
+		options -= {.info}
+	}
+
+	if .return_header in options && .return_metadata in options {
+		options -= {.return_header}
+	}
+
+	first := compress.read_u8(ctx) or_return
+	soi := cast(image.JPEG_Marker)compress.read_u8(ctx) or_return
+	if first != 0xFF && soi != .SOI {
+		return img, .Invalid_Signature
+	}
+
+	img = new(Image)
+	img.which = .JPEG
+
+	expect_EOI := false
+	huffman: [Coefficient][4]HuffmanTable
+	quantization: [4]QuantizationTable
+	color_components: [Component]ColorComponent
+	restart_interval: int
+	mcu_width: int
+	mcu_height: int
+	block_width: int
+	block_height: int
+	blocks: []Block
+	defer delete(blocks)
+
+	loop: for {
+		first = compress.read_u8(ctx) or_return
+		if first == 0xFF {
+			marker := cast(image.JPEG_Marker)compress.read_u8(ctx) or_return
+			if expect_EOI && marker != .EOI {
+				return img, .Extra_Data_After_SOS
+			}
+			#partial switch marker {
+			case cast(image.JPEG_Marker)0xFF:
+				// If we encounter multiple FF bytes then just skip them
+				continue
+			case .SOI:
+				return img, .Duplicate_SOI_Marker
+			case .APP0:
+				ident := make([dynamic]byte, 0, 16, context.temp_allocator)
+				length := cast(int)((compress.read_data(ctx, u16be) or_return) - 2)
+				for {
+					b := compress.read_u8(ctx) or_return
+					if b == 0x00 {
+						break
+					}
+					append(&ident, b)
+				}
+				if slice.equal(ident[:], image.JFIF_Magic[:]) {
+					version := compress.read_data(ctx, u16be) or_return
+					units := cast(image.JFIF_Unit)(compress.read_u8(ctx) or_return)
+					x_density := compress.read_data(ctx, u16be) or_return
+					y_density := compress.read_data(ctx, u16be) or_return
+					x_thumbnail := cast(int)compress.read_u8(ctx) or_return
+					y_thumbnail := cast(int)compress.read_u8(ctx) or_return
+					thumbnail: []image.RGB_Pixel
+
+					if x_thumbnail * y_thumbnail != 0 {
+						greyscale_thumbnail := false
+						thumbnail_size := x_thumbnail * y_thumbnail * 3
+						// According to the JFIF spec, the thumbnail should always be made of RGB pixels.
+						// But some jpegs encode single-channel thumbnails.
+						if thumbnail_size != length - 14 && thumbnail_size / 3 == length - 14 {
+							thumbnail_size = x_thumbnail * y_thumbnail
+							greyscale_thumbnail = true
+						} else {
+							return img, .Invalid_Thumbnail_Size
+						}
+						thumb_pixels := slice.reinterpret([]image.RGB_Pixel, compress.read_slice_from_memory(ctx, x_thumbnail * y_thumbnail) or_return)
+
+						if .return_metadata in options {
+							thumbnail = make([]image.RGB_Pixel, x_thumbnail * y_thumbnail)
```

Comment:
Can run out of memory, should handle the optional error from `make` here and return it as an error.